src/app/fdctl/config_parse.c

```diff
@@ -0,0 +1,358 @@
+  if( num > UINT_MAX ) {
+    FD_LOG_WARNING(( "`%s` is out of bounds (%lx)", path, num ));
+    return 0;
+  }
+  *out = (uint)num;
+  return 1;
+}
+
+static int
+fdctl_cfg_get_ushort( ushort *              out,
+                      ulong                 out_sz FD_PARAM_UNUSED,
+                      fd_pod_info_t const * info,
+                      char const *          path ) {
+  ulong num;
+  if( FD_UNLIKELY( !fdctl_cfg_get_ulong( &num, sizeof(num), info, path ) ) ) return 0;
+  if( num > USHORT_MAX ) {
+    FD_LOG_WARNING(( "`%s` is out of bounds (%lx)", path, num ));
+    return 0;
+  }
+  *out = (ushort)num;
+  return 1;
+}
+
+static int
+fdctl_cfg_get_bool( int *                 out,
+                    ulong                 out_sz FD_PARAM_UNUSED,
+                    fd_pod_info_t const * info,
+                    char const *          path ) {
+  if( FD_UNLIKELY( info->val_type != FD_POD_VAL_TYPE_INT ) ) {
+    FD_LOG_WARNING(( "invalid value for `%s`", path ));
+    return 0;
+  }
+  ulong u; fd_ulong_svw_dec( (uchar const *)info->val, &u );
+  *out = (int)u;
+  return 1;
+}
+
+/* Find leftover ******************************************************/
+
+/* fdctl_pod_find_leftover recursively searches for non-subpod keys in
+   pod.  Prints to the warning log if it finds any.  Used to detect
+   config keys that were not recognized by fdctl.  Returns 1 if a
+   leftover key was found, 0 if not.
+
+   Not thread safe (uses global buffer). */
+
+# define FDCTL_CFG_MAX_DEPTH (16)
+
+static ulong
+fdctl_pod_find_leftover_recurse( uchar *       pod,
+                                 char const ** stack,
+                                 ulong         depth ) {
+
+  if( FD_UNLIKELY( depth+1 >= FDCTL_CFG_MAX_DEPTH ) ) {
+    FD_LOG_WARNING(( "configuration file has too many nested keys" ));
+    return 0UL;
+  }
+
+  for( fd_pod_iter_t iter = fd_pod_iter_init( pod ); !fd_pod_iter_done( iter ); iter = fd_pod_iter_next( iter ) ) {
+    fd_pod_info_t info = fd_pod_iter_info( iter );
+    stack[ depth ] = info.key;
+    depth++;
+    if( FD_LIKELY( info.val_type == FD_POD_VAL_TYPE_SUBPOD ) ) {
+      ulong sub_depth = fdctl_pod_find_leftover_recurse( (uchar *)info.val, stack, depth );
+      if( FD_UNLIKELY( sub_depth ) ) return sub_depth;
+    } else {
+      return depth;
+    }
+    depth--;
+  }
+
+  return 0UL;
+}
+
+static int
+fdctl_pod_find_leftover( uchar * pod ) {
+
+  static char const * stack[ FDCTL_CFG_MAX_DEPTH ];
+  ulong depth = fdctl_pod_find_leftover_recurse( pod, stack, 0UL );
+  if( FD_LIKELY( !depth ) ) return 0;
+
+  static char path[ 64*FDCTL_CFG_MAX_DEPTH + 4 ];
+  char * c   = fd_cstr_init( path );
+  char * end = path + 64*FDCTL_CFG_MAX_DEPTH - 1;
+  for( ulong j=0UL; j<depth; j++ ) {
+    char const * key     = stack[j];
+    ulong        key_len = strlen( key );
+    if( c+key_len+1 >= end ) {
+      c = fd_cstr_append_text( c, "...", 3UL );
+      break;
+    }
+    c = fd_cstr_append_text( c, key, key_len );
```

Comment:
How do you know the key will fit in the buffer?

Reply:
There is a bounds check 3 lines above